collect_app/src/test/java/org/odk/collect/android/widgets/GeoPointWidgetTest.java

```diff
@@ -1,85 +1,217 @@
+                GeoWidgetUtils.convertCoordinatesIntoDegreeFormat(widget.getContext(), Double.parseDouble(parts[1]), "lon"),
+                GeoWidgetUtils.truncateDouble(parts[2]),
+                GeoWidgetUtils.truncateDouble(parts[3])
+        )));
+    }
+
+    @Test
+    public void whenWidgetHasAppearanceAndIsReadOnly_buttonShowsCorrectText() {
+        when(mapConfigurator.isAvailable(any())).thenReturn(true);
+        GeoPointWidget widget = createWidget(promptWithAppearance(PLACEMENT_MAP, true));
+
+        assertThat(((Button) widget.findViewById(R.id.geo_button)).getText(),
+                equalTo(widget.getContext().getString(R.string.geopoint_view_read_only)));
+    }
+
+    @Test
+    public void whenWidgetHasAppearanceAndNullAsAnswer_buttonShowsCorrectText() {
+        when(mapConfigurator.isAvailable(any())).thenReturn(true);
+        GeoPointWidget widget = createWidget(promptWithAppearanceAndAnswer(PLACEMENT_MAP, null));
+
+        assertThat(((Button) widget.findViewById(R.id.geo_button)).getText(),
+                equalTo(widget.getContext().getString(R.string.get_point)));
+    }
+
+    @Test
+    public void whenWidgetHasAppearanceAndAnswer_buttonShowsCorrectText() {
+        when(mapConfigurator.isAvailable(any())).thenReturn(true);
+        GeoPointWidget widget = createWidget(promptWithAppearanceAndAnswer(MAPS, new StringData(answer)));
+
+        assertThat(((Button) widget.findViewById(R.id.geo_button)).getText(),
+                equalTo(widget.getContext().getString(R.string.view_change_location)));
+    }
+
+    @Test
+    public void whenWidgetHasAnswer_buttonShowsCorrectText() {
+        GeoPointWidget widget = createWidget(promptWithAnswer(new StringData(answer)));
+
+        assertThat(((Button) widget.findViewById(R.id.geo_button)).getText(),
+                equalTo(widget.getContext().getString(R.string.change_location)));
+    }
+
+    @Test
+    public void whenWidgetHasNullAsAnswer_buttonShowsCorrectText() {
+        GeoPointWidget widget = createWidget(promptWithAnswer(null));
+
+        assertThat(((Button) widget.findViewById(R.id.geo_button)).getText(),
+                equalTo(widget.getContext().getString(R.string.get_point)));
+    }
+
+    @Test
+    public void buttonsShouldLaunchCorrectIntents() {
+        GeoPointWidget widget = createWidget(promptWithAnswer(null));
+        stubAllRuntimePermissionsGranted(widget, true);
+        (widget.findViewById(R.id.geo_button)).performClick();
+
+        Intent startedIntent = shadowOf(widgetTestActivity()).getNextStartedActivity();
+        Intent expectedIntent = new Intent(widgetTestActivity(), GeoPointActivity.class);
+
+        expectedIntent.putExtra(READ_ONLY, false);
+        expectedIntent.putExtra(DRAGGABLE_ONLY, false);
+        expectedIntent.putExtra(ACCURACY_THRESHOLD, DEFAULT_LOCATION_ACCURACY);
+        assertEquals(startedIntent, expectedIntent);
```

Comment:
Instead of building an `expectedIntent` here I would just assert that the `startedIntent` has the correct component and extras. It makes it easier to see what is missing when there is a fail and I'm not sure `equals` on Intent is very good. Are you familiar with how equality works in Java (`equals` vs `==` etc)?

Reply:
By building a new Intent, I might get an intent with same component name and extra values, but the `nextStartedActivity()` will not return the same intent. I think that's why the test is failing.